src/renderer.c

```diff
@@ -80,98 +102,199 @@ static const char* utf8_to_codepoint(const char *p, unsigned *dst) {
-        int source_offset = line * slot->bitmap.pitch;
-        if (font->antialiasing == FONT_ANTIALIASING_NONE) {
-          for (unsigned int column = 0; column < slot->bitmap.width; ++column) {
-            int current_source_offset = source_offset + (column / 8);
-            int source_pixel = slot->bitmap.buffer[current_source_offset];
-            pixels[++target_offset] = ((source_pixel >> (7 - (column % 8))) & 0x1) << 7;
+      if (slot->format == FT_GLYPH_FORMAT_OUTLINE
+            && font_set_style(&slot->outline, (64 / bitmaps_cached) * j, font->style)) continue;
+      if (FT_Render_Glyph(slot, render_option)) continue;
+
+      if (!slot->bitmap.width || !slot->bitmap.rows) continue;
+
+      if (slot->bitmap.pixel_mode != FT_PIXEL_MODE_LCD) {
+        unsigned int pixel_mode;
+        switch (slot->bitmap.pixel_mode) {
+          case FT_PIXEL_MODE_MONO: pixel_mode = SDL_PIXELFORMAT_INDEX1MSB; bits_per_pixel = 1; break;
+          case FT_PIXEL_MODE_GRAY: pixel_mode = SDL_PIXELFORMAT_INDEX8;    bits_per_pixel = 8; break;
+#ifdef SDL_LIL_ENDIAN
+          case FT_PIXEL_MODE_BGRA: pixel_mode = SDL_PIXELFORMAT_ARGB8888;  bits_per_pixel = 32; break;
+#else
+          case FT_PIXEL_MODE_BGRA: pixel_mode = SDL_PIXELFORMAT_BGRA8888;  bits_per_pixel = 32; break;
+#endif
+        }
+        // for BGRA surfaces we need to convert it to a surface in case we need to perform a scaled blit
+        SDL_Surface *glyph_surface = check_alloc(SDL_CreateRGBSurfaceWithFormatFrom(slot->bitmap.buffer,
+                                                                                    slot->bitmap.width, slot->bitmap.rows,
+                                                                                    bits_per_pixel, slot->bitmap.pitch,
+                                                                                    pixel_mode));
+
+        // set palettes for the indexed color modes
+        if (pixel_mode == SDL_PIXELFORMAT_INDEX1MSB)
+          SDL_SetPaletteColors(glyph_surface->format->palette, monochrome_palette, 0, array_sizeof(monochrome_palette));
+        else if (pixel_mode == SDL_PIXELFORMAT_INDEX8)
+          SDL_SetPaletteColors(glyph_surface->format->palette, grayscale_palette, 0, array_sizeof(grayscale_palette));
+
+        SDL_Rect src = { .x = 0, .y = 0, .w = slot->bitmap.width, .h = slot->bitmap.rows };
+        SDL_Rect dst = { .x = set->metrics[i].x0, .y = 0, .w = slot->bitmap.width / font->bitmap_scale, .h = slot->bitmap.rows / font->bitmap_scale };
+
+        // perform a scaled blit if necessary
+        if (font->bitmap_scale != 1.0f) {
+          if (bits_per_pixel != 32) {
+            // a scaled blit requires a 32bpp surface
+            SDL_Surface *temp = check_alloc(SDL_ConvertSurface(glyph_surface, set->surface->format, 0));
+            SDL_FreeSurface(glyph_surface);
+            glyph_surface = temp;
           }
-        } else
+
+          SDL_BlitScaled(glyph_surface, &src, set->surface, &dst);
```

Comment:
You could use `SDL_SoftStretchLinear` to have better results:
![Schermata del 2023-05-30 17-56-36](https://github.com/lite-xl/lite-xl/assets/2798487/dc0b1f10-ea9d-4f12-aea0-92086f37130a)
![Schermata del 2023-05-30 17-57-16](https://github.com/lite-xl/lite-xl/assets/2798487/569b1df6-652f-4f21-8059-97bc8739f8a6)

Still not great at small sizes tho.

Reply:
It's still an improvement.